packages/devtools/lib/src/timeline/flame_chart_canvas.dart

```diff
@@ -0,0 +1,624 @@
+
+  num _flameChartWidth;
+
+  num _zoomLevel = 1;
+
+  void paint(CanvasRenderingContext2D canvas, Rect viewport, Rect visible) {
+    // Draw the first grid item since it will have a different width than the
+    // rest.
+    final gridItem = TimelineGridNode(
+      0,
+      _flameChartInset,
+      const Duration(microseconds: 0),
+    );
+
+    gridItem.paint(canvas, viewport, visible);
+
+    num left =
+        (visible.left - _flameChartInset) ~/ currentInterval * currentInterval +
+            _flameChartInset;
+
+    while (left < visible.right) {
+      // TODO(kenzie): Instead of calculating timestamp based on position, track
+      // timestamp var and increment it by time interval represented by each
+      // grid item. See comment on https://github.com/flutter/devtools/pull/325.
+      final Duration timestamp = Duration(
+          microseconds: getTimestampForPosition(left + currentInterval));
+      final gridItem = TimelineGridNode(left, currentInterval, timestamp);
+
+      gridItem.paint(canvas, viewport, visible);
+
+      left += currentInterval;
+    }
+  }
+
+  /// Returns the timestamp rounded to the nearest microsecond for the
+  /// x-position.
+  int getTimestampForPosition(num gridItemEnd) {
+    return ((gridItemEnd - _flameChartInset) / _flameChartWidth * _duration)
+        .round();
+  }
+
+  void updateForZoom(num newZoomLevel, num newFlameChartWidth) {
+    if (_zoomLevel == newZoomLevel) {
+      return;
+    }
+
+    _flameChartWidth = newFlameChartWidth;
+
+    final log2NewZoomLevel = log2(newZoomLevel);
+
+    final gridZoomFactor = math.pow(2, log2NewZoomLevel);
+    final gridIntervalPx = baseGridInterval / gridZoomFactor;
+
+    /// The physical pixel width of the grid interval at [newZoomLevel].
+    currentInterval = gridIntervalPx * newZoomLevel;
+
+    _zoomLevel = newZoomLevel;
+  }
+}
+
+class TimelineGridNode {
+  TimelineGridNode(this.currentLeft, this.currentWidth, this.timestamp)
+      : timestampText = msText(
+          timestamp,
+          fractionDigits: timestamp.inMicroseconds == 0 ? 1 : 3,
+        );
+
+  final Duration timestamp;
+  final num currentLeft;
+  final num currentWidth;
+  final String timestampText;
+
+  static const gridLineWidth = 0.8;
+  static const gridLineColor = Color.fromRGBO(180, 180, 180, 0.5);
+  static const timestampOffsetX = 6.0;
+
+  void paint(CanvasRenderingContext2D canvas, Rect viewport, Rect visible) {
+    if (currentLeft + currentWidth < visible.left ||
+        currentLeft > visible.right) {
+      // We do not need to draw the grid node because it is out of view.
+      return;
+    }
+
+    // Draw the background for the section that will contain the timestamp. This
+    // section will be sticky to the top of the viewport. Make the background
+    // rectangles slightly larger than the current width to account for gaps
+    // that would be caused by rounding.
+    canvas.fillStyle = colorToCss(shadedBackgroundColor);
+    canvas.fillRect(
```

Comment:
why not fill the background in the paint method for the entire item instead of in this paint call? That would reduce the number of fill rect calls and avoid possible bugs with gaps.

Reply:
done.